File: sdks/typescript/src/adapters/anthropic.ts
/** Anthropic adapter (covers the @anthropic-ai/sdk client). */

import { type Usage, makeUsage } from "../schema.js";
import {
  type Path,
  type ProviderAdapter,
  type StreamAccumulator,
  read,
  readInt,
  register,
} from "./base.js";

function usageFromObj(usage: unknown): Usage {
  const inputTokens = readInt(usage, "input_tokens", 0);
  const outputTokens = readInt(usage, "output_tokens", 0);
  const cacheRead = readInt(usage, "cache_read_input_tokens", 0);
  const cacheCreation = readInt(usage, "cache_creation_input_tokens", 0);
  const reasoning = readInt(usage, "reasoning_tokens", 0);
  return makeUsage({
    // input_tokens from Anthropic already excludes cached reads; cache
    // creation is billed as (uncached) input, so fold it into input.
    input_tokens: inputTokens + cacheCreation,
    cached_input_tokens: cacheRead,
    output_tokens: outputTokens,
    reasoning_tokens: reasoning,
  });
}

class AnthropicStreamAccumulator implements StreamAccumulator {
  private usage: Usage = makeUsage();

  feed(chunk: unknown): void {
    const etype = read(chunk, "type");
    if (etype === "message_start") {
      const message = read(chunk, "message");
      const usage = read(message, "usage");
      if (usage !== null && usage !== undefined) {
        const parsed = usageFromObj(usage);
        // message_start carries input usage; keep output from deltas.
        this.usage.input_tokens = parsed.input_tokens;
        this.usage.cached_input_tokens = parsed.cached_input_tokens;
      }
    } else if (etype === "message_delta") {
      const usage = read(chunk, "usage");
      if (usage !== null && usage !== undefined) {
        this.usage.output_tokens = readInt(
          usage,
          "output_tokens",
          this.usage.output_tokens,
        );
      }
    }
    this.usage.total_tokens =
      this.usage.input_tokens +
      this.usage.output_tokens +
      this.usage.reasoning_tokens;
  }

  result(): Usage {
    return this.usage;
  }
}

class AnthropicAdapter implements ProviderAdapter {
  readonly name = "anthropic";
  readonly terminalPaths: readonly Path[] = [["messages", "create"]];

  detect(client: unknown): boolean {
    if (client === null || typeof client !== "object") return false;
    const ctorName = (client as object).constructor?.name ?? "";
    if (ctorName === "Anthropic" || ctorName === "AnthropicBedrock") return true;
    // Structural fallback: messages.create present, no chat.completions
    // (which would mark it as OpenAI-shaped).
    const messages = read(client, "messages");
    const hasMessagesCreate = typeof read(messages, "create") === "function";
    const hasChat = read(client, "chat") !== undefined;
    return hasMessagesCreate && !hasChat;
  }

  operationFor(_path: Path): string {
    return "chat";
  }

  modelFromOptions(options: Record<string, unknown>): string {
    return String(options.model ?? "");
  }

  parseUsage(response: unknown, _operation: string): Usage {
    const usage = read(response, "usage");
    if (usage === null || usage === undefined) return makeUsage();
    return usageFromObj(usage);
  }

  newStreamAccumulator(_model: string): StreamAccumulator {
    return new AnthropicStreamAccumulator();
  }

  isStream(options: Record<string, unknown>): boolean {
    return Boolean(options.stream);
  }

  prepareStreamOptions(options: Record<string, unknown>): Record<string, unknown> {
    return options;
  }
}

register(new AnthropicAdapter());
export { AnthropicAdapter };

File: sdks/python/costobs/adapters/anthropic.py
"""Anthropic adapter (covers sync + async clients)."""

from __future__ import annotations

from typing import Any, Tuple

from ..schema import Usage
from . import ProviderAdapter, StreamingUsageAccumulator, register
from ._util import get, get_int, module_root


def _usage_from_obj(usage: Any) -> Usage:
    input_tokens = get_int(usage, "input_tokens", 0)
    output_tokens = get_int(usage, "output_tokens", 0)
    cache_read = get_int(usage, "cache_read_input_tokens", 0)
    cache_creation = get_int(usage, "cache_creation_input_tokens", 0)
    # Extended thinking, if surfaced separately.
    reasoning = get_int(usage, "reasoning_tokens", 0)
    return Usage(
        # input_tokens from Anthropic already excludes cached reads; cache
        # creation is billed as (uncached) input, so fold it into input.
        input_tokens=input_tokens + cache_creation,
        cached_input_tokens=cache_read,
        output_tokens=output_tokens,
        reasoning_tokens=reasoning,
    )


class _AnthropicStreamAccumulator(StreamingUsageAccumulator):
    def feed(self, chunk: Any) -> None:
        etype = get(chunk, "type")
        if etype == "message_start":
            message = get(chunk, "message")
            usage = get(message, "usage")
            if usage is not None:
                parsed = _usage_from_obj(usage)
                # message_start carries input usage; keep output from deltas.
                self.usage.input_tokens = parsed.input_tokens
                self.usage.cached_input_tokens = parsed.cached_input_tokens
                self.usage.reasoning_tokens = parsed.reasoning_tokens
        elif etype == "message_delta":
            usage = get(chunk, "usage")
            if usage is not None:
                self.usage.output_tokens = get_int(usage, "output_tokens", self.usage.output_tokens)
                self.usage.reasoning_tokens = get_int(
                    usage, "reasoning_tokens", self.usage.reasoning_tokens
                )
        self.usage.total_tokens = (
            self.usage.input_tokens + self.usage.output_tokens + self.usage.reasoning_tokens
        )


@register
class AnthropicAdapter(ProviderAdapter):
    name = "anthropic"
    terminal_paths = (("messages", "create"),)

    def detect(self, client: Any) -> bool:
        return module_root(client) == "anthropic"

    def operation_for(self, path: Tuple[str, ...]) -> str:
        return "chat"

    def model_from_kwargs(self, kwargs: dict) -> str:
        return str(kwargs.get("model", ""))

    def parse_usage(self, response: Any, operation: str) -> Usage:
        usage = get(response, "usage")
        if usage is None:
            return Usage()
        return _usage_from_obj(usage)

    def new_stream_accumulator(self, model: str) -> StreamingUsageAccumulator:
        return _AnthropicStreamAccumulator(self, model)
